Synchronization between sourceItem and stockItem deletion for defaultSourceId

sinisa86 opened the pull request:
Synchronization between sourceItem and stockItem deletion for defaultSourceId

### Fixed Issues (if relevant)
1. magento-engcom/msi#216: Provide possibility to enable/disable synchronization logic of stock deduction in legacy tables at checkout time
2. magento-engcom/msi#218: Add Legacy Stock Synchronization at the time of Deleting StockItem (Legacy)
3. magento-engcom/msi#219: Add Update to the Legacy Inventory (StockItem) when Default Source item deleted

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

sinisa86 committed:
Implemented plugins for updating legacy CatalogInventory data on sourceItems update/delete

sinisa86 committed:
Modified synchronization plugins

sinisa86 committed:
Various fixes and API renames

bartekszymanski committed:
Merge remote-tracking branch 'remotes/origin/develop' into delete_source_items_stock_items_synchronization

bartekszymanski committed:
Fix table names

bartekszymanski committed:
Fix table names

bartekszymanski committed:
Merge remote-tracking branch 'remotes/origin/develop' into delete_source_items_stock_items_synchronization

sinisa86 committed:
Modified DeleteLegacyCatalogInventoryPlugin to update stockItem/stockStatus instead of deleting them

bartekszymanski committed:
fix products fixture

bartekszymanski committed:
Merge branch 'delete_source_items_stock_items_synchronization' of https://github.com/magento-engcom/msi into delete_source_items_stock_items_synchronization

sinisa86 committed:
Fixed Static test issue in the DeleteTest

bartekszymanski committed:
Merge remote-tracking branch 'remotes/origin/develop' into delete_source_items_stock_items_synchronization

# Conflicts:
#	app/code/Magento/InventoryApi/Test/Api/StockRepository/DeleteTest.php
#	app/code/Magento/InventoryApi/Test/_files/source_items.php

naydav commented on app/code/Magento/InventoryApi/Test/Api/StockRepository/DeleteTest.php:
```diff
@@ -0,0 +1,103 @@
+<?php
```
Pls remove this file

naydav commented on app/code/Magento/InventoryApi/Test/_files/source_items_rollback.php:
```diff
@@ -30,5 +31,8 @@
-    $sourceItemsDelete->execute($sourceItems);
+    try {
+        $sourceItemsDelete->execute($sourceItems);
+    } catch (NoSuchEntityException $e) {
```
Why do we need wrap this code in try catch? Looks like `sourceItemsDelete` does not throw `NoSuchEntityException `

naydav commented on app/code/Magento/InventoryCatalog/Model/UpdateLegacyStockItemByDefaultSourceItem.php:
```diff
@@ -0,0 +1,88 @@
+                [
+                    StockItemInterface::STOCK_ID . ' = ?' => $this->defaultStockProvider->getId(),
+                    StockItemInterface::PRODUCT_ID . ' = ?' => $productId,
+                    'website_id = ?' => 0
```
Need to check how do we write website value in this table
Looks like we need to use some service for resolving website id (instead of hardcoded value)

naydav commented on app/code/Magento/InventoryCatalog/Model/UpdateLegacyStockStatusByDefaultSourceItem.php:
```diff
@@ -0,0 +1,89 @@
+            return;
+        }
+
+        $productIds = $this->productIdLocator->retrieveProductIdsBySkus([$sourceItem->getSku()]);
```
Pls merge develop branch
We have provided global refactoring of product ids by sku resolving

Pls use `\Magento\InventoryCatalog\Model\GetProductIdsBySkusInterface`

naydav commented on app/code/Magento/InventoryCatalog/Plugin/InventoryApi/DeleteLegacyCatalogInventoryPlugin.php:
```diff
@@ -0,0 +1,62 @@
+ * Plugin help to delete related entries from the legacy catalog inventory tables cataloginventory_stock_status and
+ * cataloginventory_stock_item if deleted source item is default source item.
+ */
+class DeleteLegacyCatalogInventoryPlugin
```
Class name and description are not equal to behavior of this class

naydav commented on app/code/Magento/InventoryCatalog/Test/Integration/DeleteLegacyCatalogInventoryPluginTest.php:
```diff
@@ -0,0 +1,125 @@
+use Magento\CatalogInventory\Api\StockItemCriteriaInterfaceFactory;
+use Magento\CatalogInventory\Api\Data\StockItemCollectionInterface;
+
+class DeleteLegacyCatalogInventoryPluginTest extends TestCase
```
Change name
We does not test plugin but we test some testcase

naydav commented on app/code/Magento/InventoryCatalog/Test/Integration/DeleteLegacyCatalogInventoryPluginTest.php:
```diff
@@ -0,0 +1,125 @@
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/stocks.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/source_items.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/stock_source_link.php
+     * @magentoDbIsolation enabled
```
magentoDbIsolation is enabled by default

We have problem with this option only if we do not apply any fixtures
But in our case it is extra configuration

naydav replied:
Need to do this in all cases
It is related to the other notices

naydav commented on app/code/Magento/InventoryCatalog/Test/Integration/DeleteLegacyCatalogInventoryPluginTest.php:
```diff
@@ -0,0 +1,125 @@
+        $stockItemsBeforeDelete = $this->oldStockItemRepository->getList($criteria)->getItems();
+        $this->assertCount(1, $stockItemsBeforeDelete);
+
+        $searchCriteria = $this->searchCriteriaBuilder
```
Looks like we have a few times with code
Maybe it will be better to create sugar service (not interface but implementation) and put under hood this logic (I mean creating and configuring of Search Criteria)

naydav commented on app/code/Magento/InventoryCatalog/Test/Integration/UpdateCatalogInventoryOnSourceItemsSaveTest.php:
```diff
@@ -0,0 +1,150 @@
+        $this->sourceItemsSave->execute([$sourceItem]);
+
+        /** @var StockItemCollectionInterface $collectionBeforeChange */
+        $stockItemsAfterUpdate = $this->oldStockItemRepository->getList($criteria)->getItems();
```
Pls use `legacy` instead of `old`

naydav commented on app/code/Magento/InventoryCatalog/Test/Integration/UpdateLegacyCatalogInventoryDuringReservationPlacingTest.php:
```diff
@@ -153,4 +154,48 @@ public function testThatReservationPlacedUpdatesBothOldAndNewStocks()
+     */
+    public function testThatReservationPlacedDefersUpdatingLegacyStockWhenCanSubtractOff()
+    {
+        $this->productRepository = Bootstrap::getObjectManager()->get(ProductRepositoryInterface::class);
```
Extra line
We have productRepository  initialization in setUp

bartekszymanski committed:
Skip functional test

bartekszymanski committed:
Synchronization between sourceItem and stockItem deletion for defaultSourceId

bartekszymanski committed:
Synchronization between sourceItem and stockItem deletion for defaultSourceId

Commit:
MSI: Synchronization between sourceItem and stockItem deletion for defaultSourceId #269

-- Set to zero legacy catalocinventory data if deleted source item which is related to default source

Commit:
MSI: Synchronization between sourceItem and stockItem deletion for defaultSourceId #269

-- set data to legacy catalog inventory at source items related to default source save

Commit:
MSI: Synchronization between sourceItem and stockItem deletion for defaultSourceId #269

-- set data to legacy catalog inventory at source items related to default source save

Commit:
Merge remote-tracking branch 'origin/develop' into delete_source_items_stock_items_synchronization

Commit:
MSI: Synchronization between sourceItem and stockItem deletion for defaultSourceId #269

-- Apply inventory data changes  (qty, stock status) to legacy CatalogInventory (cataloginventory_stock_status and
   cataloginventory_stock_item tables) at the time when Reservation(-s) have been appended using MSI APIs,
   and these reservation(-s) correspond to Default Stock

Commit:
MSI: Synchronization between sourceItem and stockItem deletion for defaultSourceId #269

-- Apply inventory data changes  (qty, stock status) to legacy CatalogInventory (cataloginventory_stock_status and
   cataloginventory_stock_item tables) at the time when Reservation(-s) have been appended using MSI APIs,
   and these reservation(-s) correspond to Default Stock

Commit:
MSI: Synchronization between sourceItem and stockItem deletion for defaultSourceId #269

-- Remove legacy code

Commit:
MSI: Synchronization between sourceItem and stockItem deletion for defaultSourceId #269

-- slight refactoring

Commit:
MSI: Synchronization between sourceItem and stockItem deletion for defaultSourceId #269

-- build fixes, new task introducing

Commit:
MSI: Synchronization between sourceItem and stockItem deletion for defaultSourceId #269

-- build fixes, new task introducing
